### ISSUE TITLE
Very sluggish in Safari

### ISSUE THREAD

**Issue by alexhawkins** [jxnblk/reflexbox#46]
Not sure what has changed in 2.2.2 from 2.1.0 but my app is extremely sluggish in Safari now. When I drop back down to 2.1.0, the sluggishness disappears. I'm using reflexbox in production so hopefully you have some time to look into this. Great job so far though...reflexbox has been great for my workflow.

**Comment by jxnblk** [comment]
Thanks for reporting this! I'm testing using matchMedia listeners, and I suspect that might be what's causing the issue. You *can* disable it with the HOC, but I don't currently have a way to disable that per component or through context (though I'm planning to support that).

For now, I'd suggest locking the version to 2.1, and I'll update this thread when there's a better method to disable the listeners globally.

In my own testing, the listener performance seemed "good enough", but would you mind sharing some details about your setup? i.e. Have you set custom breakpoints? If so, how many? And, roughly how many Reflexbox components are you rendering in a single view?

**Comment by SpencerCDixon** [comment]
I've noticed a bit of sluggishness as well.  But it veery well could be due to other reasons.  One thing I noticed today was this error when resizing the window:

![screen shot 2016-09-28 at 3 16 06 pm](https://cloud.githubusercontent.com/assets/7471018/18928561/a7cc1994-858e-11e6-9a5c-5aa696522752.png)

**Comment by alexhawkins** [comment]
@jxnblk thanks. I do have custom breakpoints.  I have 9. Some of my views are pretty complex. I've up to 10 Reflexbox components in most views. A couple views have 20+. I've been using them pretty liberally, like divs. Should I not be doing that?

**Comment by nfcampos** [comment]
maybe call setState in a `requestAnimationFrame` callback in the `match` method?

**Comment by degroote22** [comment]
Tested with Chrome on Windows, ChromeOS and Android's.
I'm getting poor perfomance only on development.
I don't get console errors on production and it seems that just silencing the 'console.error's made the problem go away.

Currently I'm rendering some complicated scene nestes with Flexes and Boxes.
Resizing the window in production, for instance, gives me 100,000 console errors.

I'm using version 2.2.2 of this package.

**Comment by debrice** [comment]
I got the same `setState` error, but I would suspect they come from using material-ui at the time. There could be a conflict or a race condition between reflexbox and a third party library (like material-ui).

Also, I wanted to know @jxnblk , it could be nice to have props like hide-md (hide when medium view port or less), lg-only (show only on large viewport only)... which depending of the viewport size would return `null` instead of hiding the element through styling (to save on rendering)?

**Comment by contra** [comment]
Same setState error, here's the stack:

<img width="1390" alt="screen shot 2016-10-11 at 3 44 54 pm" src="https://cloud.githubusercontent.com/assets/425716/19285867/b7a8737c-8fc9-11e6-8434-0f945d8b5927.png">

**Comment by nfcampos** [comment]
1. the `setState` error is easy to fix, ie. in `componentWillUnmount` set `this.unmounted` to `true` and in `this.match` add `if (this.unmounted) return` to the beginning. (see PR #48)
2. after looking at this for a bit I believe the performance problem can be addressed in two ways:
  - don't create more MediaQueryList objects than we need to (see PR #47)
  - re-render only once when matchMedia listeners are called (currently each component is re-rendered as many times as there are breakpoints) this one I'm still thinking how best to fix, if anyone has any ideas do go ahead

**Comment by contra** [comment]
@nfcampos Yeah, it seems like that would be an easy fix. Also - I'm not using media queries at all with reflexbox, but it still sets up the listeners so that could be an easy perf fix as well.

**Comment by jfsiii** [comment]
It's also worth noting that Safari has severe issues with nested flexbox layouts ([webkit bug](https://bugs.webkit.org/show_bug.cgi?id=150445) & [repro case](http://codepen.io/anon/pen/zvRKYN?editors=001))

It's easy to run into this when using Flexbox for page & component layout.

**Comment by steida** [comment]
@jfsiii Isn't it fixed in Safari 10?

**Comment by jxnblk** [comment]
With v3, reflexbox uses injected CSS, so hopefully this isn't an issue any more in the new version
